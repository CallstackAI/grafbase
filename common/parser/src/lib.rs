#[cfg(test)]
#[macro_use]
extern crate assert_matches;

use std::collections::{HashMap, HashSet};

use dynaql::registry::enums::DynaqlEnums;
use dynaql::registry::scalars::{PossibleScalar, SDLDefinitionScalar};
use dynaql_parser::{parse_schema, Error as ParserError};
use quick_error::quick_error;
use rules::auth_directive::AuthDirective;
use rules::basic_type::BasicType;
use rules::check_field_lowercase::CheckFieldCamelCase;
use rules::check_known_directives::CheckAllDirectivesAreKnown;
use rules::check_type_collision::CheckTypeCollision;
use rules::check_type_validity::CheckTypeValidity;
use rules::check_types_underscore::CheckBeginsWithDoubleUnderscore;
use rules::default_directive::DefaultDirective;
use rules::default_directive_types::DefaultDirectiveTypes;
use rules::directive::Directives;
use rules::enum_type::EnumType;
use rules::extend_query_and_mutation_types::ExtendQueryAndMutationTypes;
use rules::length_directive::LengthDirective;
use rules::model_directive::ModelDirective;
use rules::one_of_directive::OneOfDirective;
use rules::openapi_directive::OpenApiVisitor;
use rules::relations::{relations_rules, RelationEngine};
use rules::resolver_directive::ResolverDirective;
use rules::search_directive::SearchDirective;
use rules::unique_directive::UniqueDirective;
use rules::unique_fields::UniqueObjectFields;
use rules::visitor::{visit, RuleError, Visitor, VisitorContext};

mod models;

use crate::rules::cache_directive::{CacheDirective, CacheVisitor};
pub use dynaql::registry::Registry;
pub use migration_detection::{required_migrations, RequiredMigration};
pub use rules::openapi_directive::{OpenApiDirective, OpenApiQueryNamingStrategy, OpenApiTransforms};

use crate::rules::scalar_hydratation::ScalarHydratation;

mod directive_de;
mod dynamic_string;
mod migration_detection;
mod registry;
mod rules;
#[cfg(test)]
mod tests;
mod utils;

quick_error! {
    #[derive(Debug)]
    pub enum Error {
        Parser(err: ParserError) {
            from()
            source(err)
            display("{}", err)
        }
        Validation(err: Vec<RuleError>) {
            from()
            display("{:?}", err)
        }
    }
}

impl Error {
    #[cfg(test)]
    fn validation_errors(self) -> Option<Vec<RuleError>> {
        if let Error::Validation(err) = self {
            Some(err)
        } else {
            None
        }
    }
}

/// Transform the input schema into a Registry
pub fn to_registry<S: AsRef<str>>(input: S) -> Result<Registry, Error> {
    Ok(to_registry_with_variables(input, &HashMap::new())?.registry)
}

pub struct ParseResult {
    pub registry: Registry,
    pub required_resolvers: HashSet<String>,
    pub openapi_directives: Vec<OpenApiDirective>,
}

/// Transform the input schema into a Registry in the context of provided environment variables
pub fn to_registry_with_variables<S: AsRef<str>>(
    input: S,
    variables: &HashMap<String, String>,
) -> Result<ParseResult, Error> {
    let directives = Directives::new()
        .with::<AuthDirective>()
        .with::<DefaultDirective>()
        .with::<LengthDirective>()
        .with::<ModelDirective>()
        .with::<OneOfDirective>()
        .with::<RelationEngine>()
        .with::<ResolverDirective>()
        .with::<UniqueDirective>()
        .with::<SearchDirective>()
        .with::<OpenApiDirective>()
        .with::<CacheDirective>();

    let mut rules = rules::visitor::VisitorNil
        .with(CheckBeginsWithDoubleUnderscore)
        .with(CheckFieldCamelCase)
        .with(CheckTypeValidity)
        .with(SearchDirective)
        .with(ModelDirective)
        .with(AuthDirective)
        .with(ResolverDirective)
        .with(CacheVisitor)
        .with(BasicType)
        .with(ExtendQueryAndMutationTypes)
        .with(EnumType)
        .with(ScalarHydratation)
        .with(LengthDirective)
        .with(UniqueObjectFields)
        .with(OpenApiVisitor)
        .with(CheckAllDirectivesAreKnown::default());

    let schema = format!(
        "{}\n{}\n{}\n{}",
        input.as_ref(),
        DynaqlEnums::sdl(),
        PossibleScalar::sdl(),
        directives.to_definition(),
    );
    let schema = parse_schema(schema)?;

    let mut ctx = VisitorContext::new_with_variables(&schema, variables);
    // Building all relations first are it requires to parse the whole schema (for ManyToMany). This allows later
    // rules to rely on RelationEngine::get to have correct information on relations.
    visit(
        &mut relations_rules().with(CheckTypeCollision::default()),
        &mut ctx,
        &schema,
    );
    if !ctx.errors.is_empty() {
        return Err(ctx.errors.into());
    }

    visit(&mut rules, &mut ctx, &schema);

    // FIXME: Get rid of the ugly double pass.
    let mut second_pass_rules = rules::visitor::VisitorNil.with(DefaultDirectiveTypes);
    visit(&mut second_pass_rules, &mut ctx, &schema);

    if !ctx.errors.is_empty() {
        return Err(ctx.errors.into());
    }

    let (registry, required_resolvers, openapi_directives) = ctx.finish();
    Ok(ParseResult {
        registry,
        required_resolvers,
        openapi_directives,
    })
}
