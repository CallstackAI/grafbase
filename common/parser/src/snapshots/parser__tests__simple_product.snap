---
source: common/parser/src/lib.rs
assertion_line: 80
expression: reg_string
---
{"types":{"Boolean":{"Scalar":{"name":"Boolean","description":"The `Boolean` scalar type represents `true` or `false`.","specified_by_url":null}},"Float":{"Scalar":{"name":"Float","description":"The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](https://en.wikipedia.org/wiki/IEEE_floating_point).","specified_by_url":null}},"ID":{"Scalar":{"name":"ID","description":null,"specified_by_url":null}},"Int":{"Scalar":{"name":"Int","description":"The `Int` scalar type represents non-fractional whole numeric values.","specified_by_url":null}},"Mutation":{"Object":{"name":"Mutation","description":null,"fields":{"productCreate":{"name":"productCreate","description":"Create a Product","args":{"input":{"name":"input","description":null,"ty":"ProductCreationInput!","default_value":null,"is_secret":false}},"ty":"ProductCreatePayload","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":"product_create_resolver","type":{"DynamoMutationResolver":{"CreateNode":{"input":{"InputTypeName":"input"},"ty":"Product"}}}},"transforms":null},"productDelete":{"name":"productDelete","description":"Delete a Product by ID","args":{"id":{"name":"id","description":null,"ty":"ID!","default_value":null,"is_secret":false}},"ty":"ProductDeletePayload","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":"product_delete_resolver","type":{"DynamoMutationResolver":{"DeleteNode":{"id":{"InputTypeName":"id"}}}}},"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"Mutation"}},"PageInfo":{"Object":{"name":"PageInfo","description":null,"fields":{"hasPreviousPage":{"name":"hasPreviousPage","description":null,"args":{},"ty":"Boolean!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":"PaginationData"}},"transforms":[{"JSONSelect":{"property":"has_previous_page","functions":[]}}]},"hasNextPage":{"name":"hasNextPage","description":null,"args":{},"ty":"Boolean!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":"PaginationData"}},"transforms":[{"JSONSelect":{"property":"has_next_page","functions":[]}}]},"startCursor":{"name":"startCursor","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":"PaginationData"}},"transforms":[{"JSONSelect":{"property":"start_cursor","functions":[]}}]},"endCursor":{"name":"endCursor","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":"PaginationData"}},"transforms":[{"JSONSelect":{"property":"end_cursor","functions":[]}}]}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"PageInfo"}},"Product":{"Object":{"name":"Product","description":null,"fields":{"id":{"name":"id","description":null,"args":{},"ty":"ID!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":{"LocalKey":{"key":"Product"}}}},"transforms":[{"DynamoSelect":{"property":"__sk"}}]},"name":{"name":"name","description":null,"args":{},"ty":"String!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":{"LocalKey":{"key":"Product"}}}},"transforms":[{"DynamoSelect":{"property":"name"}}]},"price":{"name":"price","description":"The product's price in $","args":{},"ty":"Int!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":{"LocalKey":{"key":"Product"}}}},"transforms":[{"DynamoSelect":{"property":"price"}}]}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":true,"rust_typename":"Product"}},"ProductConnection":{"Object":{"name":"ProductConnection","description":null,"fields":{"pageInfo":{"name":"pageInfo","description":"Information to aid in pagination","args":{},"ty":"PageInfo!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"edges":{"name":"edges","description":null,"args":{},"ty":"[ProductEdge]","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"ProductConnection"}},"ProductCreatePayload":{"Object":{"name":"ProductCreatePayload","description":null,"fields":{"product":{"name":"product","description":null,"args":{},"ty":"Product","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":"product_resolver","type":{"DynamoResolver":{"QueryPKSK":{"pk":{"LocalData":"id"},"sk":{"LocalData":"id"}}}}},"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"ProductCreatePayload"}},"ProductCreationInput":{"InputObject":{"name":"ProductCreationInput","description":"Input to create a new Product","input_fields":{"name":{"name":"name","description":null,"ty":"String!","default_value":null,"is_secret":false},"price":{"name":"price","description":"The product's price in $","ty":"Int!","default_value":null,"is_secret":false}},"rust_typename":"Product","oneof":false}},"ProductDeletePayload":{"Object":{"name":"ProductDeletePayload","description":null,"fields":{"deletedId":{"name":"deletedId","description":null,"args":{},"ty":"ID!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":[{"JSONSelect":{"property":"id","functions":[]}}]}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"ProductDeletePayload"}},"ProductEdge":{"Object":{"name":"ProductEdge","description":null,"fields":{"node":{"name":"node","description":null,"args":{},"ty":"Product!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"cursor":{"name":"cursor","description":null,"args":{},"ty":"String!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":null,"type":{"ContextDataResolver":{"LocalKey":{"key":"Product"}}}},"transforms":[{"DynamoSelect":{"property":"__pk"}}]}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"ProductEdge"}},"Query":{"Object":{"name":"Query","description":null,"fields":{"__schema":{"name":"__schema","description":"Access the current type schema of this server.","args":{},"ty":"__Schema!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"product":{"name":"product","description":"Get a Product by ID","args":{"id":{"name":"id","description":"Product id","ty":"ID!","default_value":null,"is_secret":false}},"ty":"Product","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":"product_resolver","type":{"DynamoResolver":{"QueryPKSK":{"pk":{"InputTypeName":"id"},"sk":{"InputTypeName":"id"}}}}},"transforms":null},"productCollection":{"name":"productCollection","description":"Paginated query to fetch the whole list of `Product`.","args":{"after":{"name":"after","description":null,"ty":"String","default_value":null,"is_secret":false},"before":{"name":"before","description":null,"ty":"String","default_value":null,"is_secret":false},"first":{"name":"first","description":null,"ty":"Int","default_value":null,"is_secret":false},"last":{"name":"last","description":null,"ty":"Int","default_value":null,"is_secret":false}},"ty":"ProductConnection","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":{"id":"product_resolver","type":{"DynamoResolver":{"ListResultByTypePaginated":{"type":{"DebugString":"Product"},"first":{"InputTypeName":"first"},"last":{"InputTypeName":"last"},"after":{"InputTypeName":"after"},"before":{"InputTypeName":"before"}}}}},"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"Query"}},"String":{"Scalar":{"name":"String","description":"The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.","specified_by_url":null}},"__Directive":{"Object":{"name":"__Directive","description":"A Directive provides a way to describe alternate runtime execution and type validation behavior in a GraphQL document.\n\nIn some cases, you need to provide options to alter GraphQL's execution behavior in ways field arguments will not suffice, such as conditionally including or skipping a field. Directives provide this by describing additional information to the executor.","fields":{"name":{"name":"name","description":null,"args":{},"ty":"String!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"description":{"name":"description","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"locations":{"name":"locations","description":null,"args":{},"ty":"[__DirectiveLocation!]!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"args":{"name":"args","description":null,"args":{},"ty":"[__InputValue!]!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"isRepeatable":{"name":"isRepeatable","description":null,"args":{},"ty":"Boolean!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"async_graphql::model::directive::__Directive"}},"__DirectiveLocation":{"Enum":{"name":"__DirectiveLocation","description":"A Directive can be adjacent to many parts of the GraphQL language, a __DirectiveLocation describes one such possible adjacencies.","enum_values":{"QUERY":{"name":"QUERY","description":"Location adjacent to a query operation.","deprecation":"NoDeprecated"},"MUTATION":{"name":"MUTATION","description":"Location adjacent to a mutation operation.","deprecation":"NoDeprecated"},"SUBSCRIPTION":{"name":"SUBSCRIPTION","description":"Location adjacent to a subscription operation.","deprecation":"NoDeprecated"},"FIELD":{"name":"FIELD","description":"Location adjacent to a field.","deprecation":"NoDeprecated"},"FRAGMENT_DEFINITION":{"name":"FRAGMENT_DEFINITION","description":"Location adjacent to a fragment definition.","deprecation":"NoDeprecated"},"FRAGMENT_SPREAD":{"name":"FRAGMENT_SPREAD","description":"Location adjacent to a fragment spread.","deprecation":"NoDeprecated"},"INLINE_FRAGMENT":{"name":"INLINE_FRAGMENT","description":"Location adjacent to an inline fragment.","deprecation":"NoDeprecated"},"VARIABLE_DEFINITION":{"name":"VARIABLE_DEFINITION","description":"Location adjacent to a variable definition.","deprecation":"NoDeprecated"},"SCHEMA":{"name":"SCHEMA","description":"Location adjacent to a schema definition.","deprecation":"NoDeprecated"},"SCALAR":{"name":"SCALAR","description":"Location adjacent to a scalar definition.","deprecation":"NoDeprecated"},"OBJECT":{"name":"OBJECT","description":"Location adjacent to an object type definition.","deprecation":"NoDeprecated"},"FIELD_DEFINITION":{"name":"FIELD_DEFINITION","description":"Location adjacent to a field definition.","deprecation":"NoDeprecated"},"ARGUMENT_DEFINITION":{"name":"ARGUMENT_DEFINITION","description":"Location adjacent to an argument definition.","deprecation":"NoDeprecated"},"INTERFACE":{"name":"INTERFACE","description":"Location adjacent to an interface definition.","deprecation":"NoDeprecated"},"UNION":{"name":"UNION","description":"Location adjacent to a union definition.","deprecation":"NoDeprecated"},"ENUM":{"name":"ENUM","description":"Location adjacent to an enum definition.","deprecation":"NoDeprecated"},"ENUM_VALUE":{"name":"ENUM_VALUE","description":"Location adjacent to an enum value definition.","deprecation":"NoDeprecated"},"INPUT_OBJECT":{"name":"INPUT_OBJECT","description":"Location adjacent to an input object type definition.","deprecation":"NoDeprecated"},"INPUT_FIELD_DEFINITION":{"name":"INPUT_FIELD_DEFINITION","description":"Location adjacent to an input object field definition.","deprecation":"NoDeprecated"}},"rust_typename":"async_graphql::model::directive::__DirectiveLocation"}},"__EnumValue":{"Object":{"name":"__EnumValue","description":"One possible value for a given Enum. Enum values are unique values, not a placeholder for a string or numeric value. However an Enum value is returned in a JSON response as a string.","fields":{"name":{"name":"name","description":null,"args":{},"ty":"String!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"description":{"name":"description","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"isDeprecated":{"name":"isDeprecated","description":null,"args":{},"ty":"Boolean!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"deprecationReason":{"name":"deprecationReason","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"async_graphql::model::enum_value::__EnumValue"}},"__Field":{"Object":{"name":"__Field","description":"Object and Interface types are described by a list of Fields, each of which has a name, potentially a list of arguments, and a return type.","fields":{"name":{"name":"name","description":null,"args":{},"ty":"String!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"description":{"name":"description","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"args":{"name":"args","description":null,"args":{},"ty":"[__InputValue!]!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"type":{"name":"type","description":null,"args":{},"ty":"__Type!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"isDeprecated":{"name":"isDeprecated","description":null,"args":{},"ty":"Boolean!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"deprecationReason":{"name":"deprecationReason","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"async_graphql::model::field::__Field"}},"__InputValue":{"Object":{"name":"__InputValue","description":"Arguments provided to Fields or Directives and the input fields of an InputObject are represented as Input Values which describe their type and optionally a default value.","fields":{"name":{"name":"name","description":null,"args":{},"ty":"String!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"description":{"name":"description","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"type":{"name":"type","description":null,"args":{},"ty":"__Type!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"defaultValue":{"name":"defaultValue","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"async_graphql::model::input_value::__InputValue"}},"__Schema":{"Object":{"name":"__Schema","description":"A GraphQL Schema defines the capabilities of a GraphQL server. It exposes all available types and directives on the server, as well as the entry points for query, mutation, and subscription operations.","fields":{"types":{"name":"types","description":"A list of all types supported by this server.","args":{},"ty":"[__Type!]!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"queryType":{"name":"queryType","description":"The type that query operations will be rooted at.","args":{},"ty":"__Type!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"mutationType":{"name":"mutationType","description":"If this server supports mutation, the type that mutation operations will be rooted at.","args":{},"ty":"__Type","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"subscriptionType":{"name":"subscriptionType","description":"If this server support subscription, the type that subscription operations will be rooted at.","args":{},"ty":"__Type","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"directives":{"name":"directives","description":"A list of all directives supported by this server.","args":{},"ty":"[__Directive!]!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"async_graphql::model::schema::__Schema"}},"__Type":{"Object":{"name":"__Type","description":"The fundamental unit of any GraphQL Schema is the type. There are many kinds of types in GraphQL as represented by the `__TypeKind` enum.\n\nDepending on the kind of a type, certain fields describe information about that type. Scalar types provide no information beyond a name and description, while Enum types provide their values. Object and Interface types provide the fields they describe. Abstract types, Union and Interface, provide the Object types possible at runtime. List and NonNull types compose other types.","fields":{"kind":{"name":"kind","description":null,"args":{},"ty":"__TypeKind!","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"name":{"name":"name","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"description":{"name":"description","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"fields":{"name":"fields","description":null,"args":{"includeDeprecated":{"name":"includeDeprecated","description":null,"ty":"Boolean!","default_value":"false","is_secret":false}},"ty":"[__Field!]","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"interfaces":{"name":"interfaces","description":null,"args":{},"ty":"[__Type!]","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"possibleTypes":{"name":"possibleTypes","description":null,"args":{},"ty":"[__Type!]","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"enumValues":{"name":"enumValues","description":null,"args":{"includeDeprecated":{"name":"includeDeprecated","description":null,"ty":"Boolean!","default_value":"false","is_secret":false}},"ty":"[__EnumValue!]","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"inputFields":{"name":"inputFields","description":null,"args":{},"ty":"[__InputValue!]","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"ofType":{"name":"ofType","description":null,"args":{},"ty":"__Type","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null},"specifiedByURL":{"name":"specifiedByURL","description":null,"args":{},"ty":"String","deprecation":"NoDeprecated","cache_control":{"public":true,"max_age":0},"external":false,"requires":null,"provides":null,"edges":[],"relation":null,"resolve":null,"transforms":null}},"cache_control":{"public":true,"max_age":0},"extends":false,"keys":null,"is_subscription":false,"is_node":false,"rust_typename":"async_graphql::model::type::__Type"}},"__TypeKind":{"Enum":{"name":"__TypeKind","description":"An enum describing what kind of type a given `__Type` is.","enum_values":{"SCALAR":{"name":"SCALAR","description":"Indicates this type is a scalar.","deprecation":"NoDeprecated"},"OBJECT":{"name":"OBJECT","description":"Indicates this type is an object. `fields` and `interfaces` are valid fields.","deprecation":"NoDeprecated"},"INTERFACE":{"name":"INTERFACE","description":"Indicates this type is an interface. `fields` and `possibleTypes` are valid fields.","deprecation":"NoDeprecated"},"UNION":{"name":"UNION","description":"Indicates this type is a union. `possibleTypes` is a valid field.","deprecation":"NoDeprecated"},"ENUM":{"name":"ENUM","description":"Indicates this type is an enum. `enumValues` is a valid field.","deprecation":"NoDeprecated"},"INPUT_OBJECT":{"name":"INPUT_OBJECT","description":"Indicates this type is an input object. `inputFields` is a valid field.","deprecation":"NoDeprecated"},"LIST":{"name":"LIST","description":"Indicates this type is a list. `ofType` is a valid field.","deprecation":"NoDeprecated"},"NON_NULL":{"name":"NON_NULL","description":"Indicates this type is a non-null. `ofType` is a valid field.","deprecation":"NoDeprecated"}},"rust_typename":"async_graphql::model::kind::__TypeKind"}}},"directives":{"skip":{"name":"skip","description":"Directs the executor to skip this field or fragment when the `if` argument is true.","locations":["FIELD","FRAGMENT_SPREAD","INLINE_FRAGMENT"],"args":{"if":{"name":"if","description":"Skipped when true.","ty":"Boolean!","default_value":null,"is_secret":false}},"is_repeatable":false},"include":{"name":"include","description":"Directs the executor to include this field or fragment only when the `if` argument is true.","locations":["FIELD","FRAGMENT_SPREAD","INLINE_FRAGMENT"],"args":{"if":{"name":"if","description":"Included when true.","ty":"Boolean!","default_value":null,"is_secret":false}},"is_repeatable":false}},"implements":{},"query_type":"Query","mutation_type":"Mutation","subscription_type":null,"disable_introspection":false,"enable_federation":false,"federation_subscription":false}
